### PR TITLE
[zkpp] Add new port

### PR DESCRIPTION
**Pull request by SamuelMarks**
> [Apache ZooKeeper](https://zookeeper.apache.org) is a centralized service for maintaining configuration information, naming, providing distributed synchronization, and providing group services.

Depends on: https://github.com/microsoft/vcpkg/pull/7000

### CODE REVIEW

**Comment by grdowns** [comment]
Hey @SamuelMarks, thanks for the new port! #7000 has been merged which means useful CI results for this PR are now available. If you have the time, please take a look at the failing builds and attempt to fix them. Otherwise I'll be able to take a look when I get around to it. Thanks again!

**Comment by Voskrese** [comment]
CONTROL

~~Build-Depends: zookeeper-mt~~
work
Build-Depends: zookeeper

**Comment by Voskrese** [comment]
`cl : Command line warning D9002 : ignoring unknown option '--std=c++17'`

msvc `/std:c++17 /permissive-`

**Comment by Voskrese** [comment]
`zkpp\src\v0.2.3-997d0d35f7\src\zk\error.cpp(6): fatal error C1083: Cannot open include file: 'zookeeper/zookeeper.h': No such file or directory`

**Comment by raid-7** [comment]
[Zookeeper library port](https://github.com/microsoft/vcpkg/tree/master/ports/zookeeper) does not correctly expose include directories. That's why this pull request cannot be built. I have a small patch to fix the issue.

```
diff --git a/ports/zookeeper/cmake.patch b/ports/zookeeper/cmake.patch
index 4dd8ef93b..8ae1e24da 100644
--- a/ports/zookeeper/cmake.patch
+++ b/ports/zookeeper/cmake.patch
@@ -19 +19 @@ index 24a5a1b..1b0ce4a 100644
-+target_include_directories(zookeeper PUBLIC
++target_include_directories(zookeeper INTERFACE "${CMAKE_INSTALL_PREFIX}/include" PUBLIC
```

Is it suitable to make commit to this branch or a separated pull request has to be created?

**Comment by SamuelMarks** [comment]
@raid-7 Commit to this branch… that will update the builds

Thanks

File: ports/zkpp/CONTROL [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Source: zkpp
+Version: 0.2.3-3
```
Comment (NancyLi1013) [inline ports/zkpp/CONTROL]: Could you update the version as `0.2.3`?

File: ports/zkpp/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,28 @@
+include(vcpkg_common_functions)
```
Comment (NancyLi1013) [inline ports/zkpp/portfile.cmake]: `include(vcpkg_common_functions)` is no longer needed.
Could you please remove this line?

File: ports/zkpp/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+)
+
+vcpkg_install_cmake()
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
+file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/zkpp RENAME copyright)
```
Comment (NancyLi1013) [inline ports/zkpp/portfile.cmake]: Could you update 
`file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/zkpp RENAME copyright)`
as `file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)`?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
The failures on windows are like this:
```
cl : Command line warning D9002 : ignoring unknown option '--std=c++17'
class template variant is only available with C++17 or later.
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\acl.hpp(118): warning C5030: attribute 'gnu::pure' is not recognized
class template optional is only available with C++17 or later.
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(17): error C2039: 'optional': is not a member of 'std'
C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\include\future(1653): note: see declaration of 'std'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(17): error C2061: syntax error: identifier 'optional'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(19): error C2039: 'nullopt_t': is not a member of 'std'
C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\include\future(1653): note: see declaration of 'std'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(21): error C2039: 'nullopt': is not a member of 'std'
C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\include\future(1653): note: see declaration of 'std'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(21): error C2873: 'nullopt': symbol cannot be used in a using-declaration
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(25): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(25): error C2988: unrecognizable template declaration/definition
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(25): error C2143: syntax error: missing ',' before '<'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(25): error C3613: missing return type after '->' ('int' assumed)
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(25): error C3646: 'optional': unknown override specifier
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(25): error C2988: unrecognizable template declaration/definition
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(25): error C2059: syntax error: '<'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(26): error C2143: syntax error: missing ';' before '{'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(26): error C2447: '{': missing function header (old-style formal list?)
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(34): error C2988: unrecognizable template declaration/definition
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(34): error C2143: syntax error: missing ';' before '<'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(34): error C2059: syntax error: '<'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(35): error C2143: syntax error: missing ';' before '{'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\optional.hpp(35): error C2447: '{': missing function header (old-style formal list?)
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\types.hpp(29): error C2631: 'strong_id': a class or enum cannot be defined in an alias template
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\multi.cpp: error C2988: unrecognizable template declaration/definition
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\multi.cpp: error C2059: syntax error: 'namespace'
C:\vsts\_work\4\s\buildtrees\zkpp\src\v0.2.3-997d0d35f7\src\zk\multi.cpp: fatal error C1903: unable to recover from previous error(s); stopping compilation
```
Could you please look into it? 
I also noticed that you updated the patch in `zookeeper`. So could you also bump the version in `ports/zookeeper/CONTROL`?

**Comment by raid-7** [comment]
The upstream library is claimed to be Linux-only, though it also works on OSX. So this port is not intended to support Windows.

**Comment by NancyLi1013** [comment]
Thanks for your info.
Please` add vcpkg_fail_port_install(ON_TARGET "Windows")` to the top of portfile.cmake.
Also update the following codes to [ci.baseline.txt](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt).
```
zkpp:x86-windows=fail
zkpp:x64-windows=fail
zkpp:x64-windows-static=fail
zkpp:arm64-windows=fail
zkpp:x64-uwp=fail
zkpp:arm-uwp=fail
```

**Review by NancyLi1013 — APPROVED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by SamuelMarks** [comment]
@NancyLi1013 Thanks, can we merge this to master now? :)

**Comment by NancyLi1013** [comment]
Hi @SamuelMarks 

LGTM now. But we also need to follow the process to handle all PRs if there is no high priority.

Thanks again for your PR and patience.

**Comment by strega-nil** [comment]
Is there a reason that you've also updated zookeeper in this PR? Like, do there need to be updates so that zkpp works, or are they just nice updates that can be done in their own PR?

**Comment by raid-7** [comment]
This port relies on zookeeper. There was a critical issue in zookeeper port, which made it actually unusable.

In more details, patched zookeeper CMakeLists.txt utilized `CMAKE_INSTALL_INCLUDEDIR` variable but didn't import `GNUInstallDirs`. Consequently, it didn't correctly expose include directories. Therefore zkpp failed to be built just because `zookeeper.h` couldn't be found.

**Comment by strega-nil** [comment]
Ah, cool, thanks :)

**Comment by strega-nil** [comment]
My second question is -- I just tried to use zookeeper by itself on macOS, and it seems to require `Threads` if one depends on it; what's going on there, do you know?

**Comment by raid-7** [comment]
Asynchronous API of the library uses threads to perform some internal work. To be honest, I don't know more details. You can take a look at [source code](https://github.com/apache/zookeeper/tree/branch-3.5.5/zookeeper-client/zookeeper-client-c/src). Particularly, threads are created in [this file](https://github.com/apache/zookeeper/blob/branch-3.5.5/zookeeper-client/zookeeper-client-c/src/mt_adaptor.c).

**Comment by strega-nil** [comment]
@raid-7 yeah, that latter bit was my question.

**Comment by SamuelMarks** [comment]
Great, so have we resolved everything required for this to be merged?

**Comment by JackBoosY** [comment]
@strega-nil Did you mean that this PR needs to fix the `Threads` issue?

**Comment by strega-nil** [comment]
No, sorry, I just forgot to hit merge -.-

**Comment by strega-nil** [comment]
Thanks so much @SamuelMarks :D

**Pull request merged**